doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,252 @@
+    target = node["reftarget"]
+    location = get_location(node, app)
+
+    dtype = "{}:{}".format(domain, typ)
+
+    record[(dtype, target)].add(location)
+
+
+def record_missing_reference_handler(app, env, node, contnode):
+    """
+    When the sphinx app notices a missing reference, it emits an
+    event which calls this function. This function records the missing
+    references for analysis at the end of the sphinx build.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    if not hasattr(env, "missing_reference_events"):
+        env.missing_reference_events = defaultdict(set)
+
+    record_missing_reference(app, env.missing_reference_events, node)
+
+
+def get_location(node, app):
+    """
+    Given a docutils node and a sphinx application, return a string
+    representation of the source location of this node.
+
+    Usually, this will be of the form "path/to/file:linenumber". Two
+    special values can be emitted, "<external>" for paths which are
+    not contained in this source tree (e.g. docstrings included from
+    other modules) or "<unknown>", inidcating that the sphinx application
+    cannot locate the original source file (usually because an extension
+    has injected text into the sphinx parsing engine).
+    """
+    (path, line) = get_source_line(node)
+
+    if path:
+
+        basepath = os.path.abspath(os.path.join(app.confdir, ".."))
+        path = os.path.relpath(path, start=basepath)
+
+        if path.startswith(os.path.pardir):
+            path = os.path.join("<external>", os.path.basename(path))
+
+    else:
+        path = "<unknown>"
+
+    if line:
+        line = str(line)
+    else:
+        line = ""
+
+    return "%s:%s" % (path, line)
+
+
+def save_missing_references_handler(app, exc):
+    """
+    At the end of the sphinx build, either save the missing references to a
+    JSON file. Also ensure that all lines of the existing JSON file are still
+    necessary.
+    """
+    if not app.config.missing_references_enabled:
+        # no-op when we are disabled.
+        return
+
+    json_path = os.path.join(app.confdir,
+                             app.config.missing_references_filename)
+
+    references_warnings = getattr(app.env, 'missing_reference_warnings', {})
+
+    # This is a dictionary of {(dtype,target): locations}
+    references_ignored = getattr(app.env,
+        'missing_references_ignored_references', {})
+    references_events = getattr(app.env, 'missing_reference_events', {})
+
+    # Warn about any reference which is no longer missing.
+    for (dtype, target), locations in references_ignored.items():
+        missing_reference_locations = references_events.get(
+                (dtype, target), [])
+
+        # For each ignored reference location, ensure a missing reference was
+        # observed. If it wasn't observed, issue a warning.
+        for ignored_refernece_location in locations:
```

Comment:
typo? reference (funny that you typoed it multiple times :p)

Reply:
The curse of an IDE: it will help me keep consistent typos!